flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/PulsarWriter.java

```diff
@@ -0,0 +1,311 @@
+        }
+    }
+
+    // ------------------------------internal method------------------------------
+
+    private synchronized PulsarClientImpl getPulsarClient() {
+        if (pulsarClient != null) {
+            return pulsarClient;
+        }
+        pulsarClient = (PulsarClientImpl) PulsarConfigUtils.createClient(configuration);
+        closer.register(pulsarClient);
+        return pulsarClient;
+    }
+
+    protected Producer<?> getProducer(String topic) throws PulsarClientException {
+        Producer<?> producer = topic2Producer.get(topic);
+        if (producer != null && producer.isConnected()) {
+            return producer;
+        }
+
+        final ProducerBuilder<?> producerBuilder =
+                PulsarSinkConfigUtils.createProducerBuilder(
+                        getPulsarClient(), serializationSchema.getSchema(), configuration);
+        producerBuilder.topic(topic);
+
+        if (partitionSelector != null) {
+            producerBuilder.messageRouter(generateMessageRouter());
+        }
+        producer = producerBuilder.create();
+        closer.register(producer);
+        uploadSchema(topic);
+        topic2Producer.put(topic, producer);
+        return producer;
+    }
+
+    @SuppressWarnings("unchecked")
+    private MessageRouter generateMessageRouter() {
+        return new MessageRouter() {
+            @Override
+            public int choosePartition(Message<?> msg, TopicMetadata metadata) {
+                return partitionSelector.select(
+                        (Message<IN>) msg,
+                        new org.apache.flink.connector.pulsar.sink.writer.selector.TopicMetadata(
+                                metadata.numPartitions()));
+            }
+        };
+    }
+
+    protected BiConsumer<MessageId, Throwable> initializeSendCallback(
+            SinkConfiguration sinkConfiguration, MailboxExecutor mailboxExecutor) {
+        if (sinkConfiguration.isFailOnWrite()) {
+            return (t, u) -> {
+                if (u == null) {
+                    acknowledgeMessage();
+                    return;
+                }
+                mailboxExecutor.execute(
+                        () -> {
+                            throw new FlinkRuntimeException(u);
+                        },
+                        "Failed to send data to Pulsar");
+            };
+        } else {
+            return (t, u) -> {
+                if (u != null) {
+                    LOG.error(
+                            "Error while sending message to Pulsar: {}",
+                            ExceptionUtils.stringifyException(u));
+                }
+                acknowledgeMessage();
+            };
+        }
+    }
+
+    private void acknowledgeMessage() {
+        synchronized (pendingRecords) {
+            if (pendingRecords.decrementAndGet() == 0L) {
+                pendingRecords.notifyAll();
+            }
+        }
+    }
```

Comment:
Can you explain this block? In general, synchronization should not be necessary if everything is executed in the mailbox.